src/dbnode/storage/database.go

```diff
@@ -694,17 +696,31 @@ func (d *db) writeBatch(
 			errHandler.HandleError(write.OriginalIndex, err)
 		}
 
-		// Need to set the outcome in the success case so the commitlog gets the updated
-		// series object which contains identifiers (like the series ID) whose lifecycle
-		// live longer than the span of this request, making them safe for use by the async
-		// commitlog. Need to set the outcome in the error case so that the commitlog knows
-		// to skip this entry.
-		writes.SetOutcome(i, series, err)
+		if !shouldWrite {
+			// This series has no additional information that needs to be written to
+			// the commit log; set this series to skip writing to the commit log.
+			writes.SetSkipWrite(i)
+			numSkipped++
+		} else {
+			// Need to set the outcome in the success case so the commitlog gets the
+			// updated series object which contains identifiers (like the series ID)
+			// whose lifecycle lives longer than the span of this request, making them
+			// safe for use by the async commitlog. Need to set the outcome in the
+			// error case so that the commitlog knows to skip this entry.
+			writes.SetOutcome(i, series, err)
+		}
+	}
+
+	if numSkipped > 0 {
+		// Sanitize the writes; if there are any skipping writes, they get dropped
+		// here. Ensure that there are remaining non-errored writes that need to go
+		// through remaining.
+		writes.Sanitize()
```

Comment:
I don't like the idea of reshuffling this entire slice on the hot path (also, it might be fine, but I worry it might interfere with some of the error handling logic in the tchannel layer).

I think the commitlog itself already checks if the write had an error and if so skips it. Instead of sanitizing, could we just add a new method for each write so that the commitlog can do something like:

```golang
if write.ShouldWrite() {

}
```

and then

```golang
func (w *write) ShouldWrite() bool {
    return w.SkipWrite == False && w.Err == nil
}
```

Reply:
Sure, sounds good; I was thinking that the reslice was cheaper than having these series go through the entire commitlog write path only to skip at the write step itself.